#Decision_Tree

1. What is a Decision Tree, and how does it work in the context of
classification?
   - A Decision Tree is a flowchart-like structure where each internal node represents a "test" on an attribute, each branch represents the outcome of the test, and each leaf node represents a class label.
     
     In the context of classification, a Decision Tree works by recursively partitioning the data based on the values of the features. The goal is to create partitions that are as "pure" as possible, meaning that the instances within each partition belong to the same class.

     The process starts at the root node with the entire dataset. The algorithm selects the best feature to split the data based on a criterion like Gini impurity or information gain, which measure how well the split separates the classes.

     This process is repeated at each child node until a stopping criterion is met, such as reaching a maximum depth, having a minimum number of instances in a leaf node, or achieving a certain level of purity.

     To classify a new instance, you traverse the tree from the root, following the branches that correspond to the instance's feature values. When you reach a leaf node, the class label associated with that node is the predicted class for the instance.

2. Explain the concepts of Gini Impurity and Entropy as impurity measures.
How do they impact the splits in a Decision Tree?
   - Gini Impurity: This measures the probability of incorrectly classifying a randomly chosen element in the dataset if it were randomly labeled according to the distribution of labels in the subset. A Gini impurity of 0 means the subset is perfectly pure, while a Gini impurity of 1 means the subset is completely impure.
   - Entropy: This measures the amount of uncertainty or randomness in the data. Like Gini impurity, an entropy of 0 means perfect purity, and higher entropy indicates greater impurity.

3. What is the difference between Pre-Pruning and Post-Pruning in Decision
Trees? Give one practical advantage of using each.
   - Pre-pruning and post-pruning are two techniques used to prevent overfitting in Decision Trees. Overfitting occurs when the tree becomes too complex and learns the training data too well, resulting in poor performance on unseen data.
   
   - Pre-Pruning: This technique stops the tree building process early. It sets criteria before the tree is fully grown, such as a maximum depth, minimum number of samples required to split an internal node, or minimum number of samples in a leaf node. If a split does not meet these criteria, the node is not split further, and it becomes a leaf node.
   
   - Pre-Pruning: This technique stops the tree building process early. It sets criteria before the tree is fully grown, such as a maximum depth, minimum number of samples required to split an internal node, or minimum number of samples in a leaf node. If a split does not meet these criteria, the node is not split further, and it becomes a leaf node.

4. What is Information Gain in Decision Trees, and why is it important for
choosing the best split?
   - Information Gain is a key concept in Decision Trees used to determine the effectiveness of a split.

   Here's why it's important:
   - Information Gain is important for choosing the best split because it helps the Decision Tree algorithm select the feature and split point that best separates the data into distinct classes.
   - A higher information gain indicates a more effective split, as it results in a greater reduction in uncertainty and leads to more homogeneous child nodes.
   - The algorithm will choose the split that maximizes information gain at each step, recursively building the tree in a way that efficiently partitions the data based on the most informative features.

5. What are some common real-world applications of Decision Trees, and
what are their main advantages and limitations?
   - Medical Diagnosis: Decision Trees can be used to build models that help diagnose diseases based on patient symptoms and medical history.
   - Credit Risk Assessment: Financial institutions use Decision Trees to assess the creditworthiness of loan applicants based on factors like income, credit score, and employment history.
   - Spam Filtering: Email providers can use Decision Trees to classify emails as spam or not spam based on the content and characteristics of the email.
   - Customer Relationship Management (CRM): Businesses use Decision Trees to analyze customer data and predict customer behavior, such as churn risk or likelihood of purchasing a product.
   - Fraud Detection: Decision Trees can be employed to identify fraudulent transactions in various industries, including banking and e-commerce.
   - Image Classification: In some cases, Decision Trees can be used for classifying images based on their features.

   Advantages:
   - Decision Trees are intuitive and easy to visualize, making it simple to understand how the model makes predictions.
   - They can work with both types of data without requiring extensive preprocessing.
   - They can capture complex, non-linear relationships between features and the target variable.
   - Compared to some other algorithms, Decision Trees require less data cleaning and normalization.

   Diadvantages:
   - Without proper pruning, Decision Trees can easily overfit the training data, leading to poor performance on unseen data.
   - Small changes in the data can lead to significant changes in the structure of the tree.
   - Decision Trees can be biased towards features with a larger number of levels or categories.
   - If the training data is imbalanced, the Decision Tree may create biased trees that favor the majority class.

6. Write a Python program to:

● Load the Iris Dataset

● Train a Decision Tree Classifier using the Gini criterion

● Print the model’s accuracy and feature importances

In [2]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score

iris = load_iris()
X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

clf = DecisionTreeClassifier(criterion='gini', random_state=42)
clf.fit(X_train, y_train)

y_pred = clf.predict(X_test)

accuracy = accuracy_score(y_test, y_pred)
print(f"Model Accuracy: {accuracy:.2f}")

print("\nFeature Importances:")
for feature, importance in zip(iris.feature_names, clf.feature_importances_):
    print(f"{feature}: {importance:.4f}")

Model Accuracy: 1.00

Feature Importances:
sepal length (cm): 0.0000
sepal width (cm): 0.0191
petal length (cm): 0.8933
petal width (cm): 0.0876


7. Write a Python program to:

● Load the Iris Dataset

● Train a Decision Tree Classifier with max_depth=3 and compare its accuracy to
a fully-grown tree.

In [7]:
from sklearn.datasets import load_iris # Import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split # Import train_test_split
from sklearn.metrics import accuracy_score

# Load the Iris dataset
iris = load_iris() # Load Iris dataset
X = iris.data
y = iris.target

# Split the dataset into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42) # Split Iris data

# Train a Decision Tree Classifier with max_depth=3
clf_pruned = DecisionTreeClassifier(criterion='gini', max_depth=3, random_state=42)
clf_pruned.fit(X_train, y_train)

# Make predictions on the test set with the pruned tree
y_pred_pruned = clf_pruned.predict(X_test)

# Calculate and print the accuracy of the pruned tree
accuracy_pruned = accuracy_score(y_test, y_pred_pruned)
print(f"Model Accuracy (max_depth=3): {accuracy_pruned:.2f}")

Model Accuracy (max_depth=3): 1.00


Comparing the accuracy of the fully-grown tree (which was {{accuracy:.2f}}) to the tree with `max_depth=3` (which is {{accuracy_pruned:.2f}}), we can see how limiting the depth of the tree impacts its performance on the test set.

8. Write a Python program to:

● Load the California Housing dataset from sklearn

● Train a Decision Tree Regressor

● Print the Mean Squared Error (MSE) and feature importances

In [5]:
from sklearn.datasets import fetch_california_housing
from sklearn.tree import DecisionTreeRegressor
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_squared_error

california_housing = fetch_california_housing()
X = california_housing.data
y = california_housing.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

regressor = DecisionTreeRegressor(random_state=42)
regressor.fit(X_train, y_train)

y_pred = regressor.predict(X_test)

mse = mean_squared_error(y_test, y_pred)
print(f"Mean Squared Error (MSE): {mse:.2f}")

print("\nFeature Importances:")
for feature, importance in zip(california_housing.feature_names, regressor.feature_importances_):
    print(f"{feature}: {importance:.4f}")

Mean Squared Error (MSE): 0.53

Feature Importances:
MedInc: 0.5235
HouseAge: 0.0521
AveRooms: 0.0494
AveBedrms: 0.0250
Population: 0.0322
AveOccup: 0.1390
Latitude: 0.0900
Longitude: 0.0888


9. Write a Python program to:

● Load the Iris Dataset

● Tune the Decision Tree’s max_depth and min_samples_split using
GridSearchCV

● Print the best parameters and the resulting model accuracy

In [9]:
from sklearn.datasets import load_iris
from sklearn.tree import DecisionTreeClassifier
from sklearn.model_selection import train_test_split, GridSearchCV
from sklearn.metrics import accuracy_score

iris = load_iris()
X = iris.data
y = iris.target

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.3, random_state=42)

param_grid = {
    'max_depth': [None, 2, 3, 4, 5, 10],
    'min_samples_split': [2, 5, 10, 20]
}

clf = DecisionTreeClassifier(random_state=42)

grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy')

grid_search.fit(X_train, y_train)

print("Best parameters found by GridSearchCV:")
print(grid_search.best_params_)

best_clf = grid_search.best_estimator_

y_pred_best = best_clf.predict(X_test)

accuracy_best = accuracy_score(y_test, y_pred_best)
print(f"\nAccuracy of the best model: {accuracy_best:.2f}")

Best parameters found by GridSearchCV:
{'max_depth': None, 'min_samples_split': 10}

Accuracy of the best model: 1.00


10. : Imagine you’re working as a data scientist for a healthcare company that wants to predict whether a patient has a certain disease. You have a large dataset with mixed data types and some missing values.

Explain the step-by-step process you would follow to:

● Handle the missing values

● Encode the categorical features

● Train a Decision Tree model

● Tune its hyperparameters

● Evaluate its performance
And describe what business value this model could provide in the real-world setting.

# Task
Explain the step-by-step process to handle missing values, encode categorical features, train and tune a Decision Tree model, and evaluate its performance for predicting a disease in a healthcare dataset with mixed data types and missing values. Describe the business value of this model.

## Data loading and initial exploration

### Subtask:
Load the dataset and perform initial exploration to understand its structure, data types, and identify missing values and categorical features.


**Reasoning**:
Load the dataset into a pandas DataFrame and display the first few rows to understand its structure and content. Then, check the data types of each column and calculate the number and percentage of missing values to identify missing values and categorical features.



In [18]:
import pandas as pd

df = pd.DataFrame(data)

display(df.head())

display(df.info())

missing_values = df.isnull().sum()
display(missing_values)

missing_percentage = (missing_values / len(df)) * 100
display(missing_percentage)

PatientID  Age  Gender  BloodPressure  Cholesterol   Outcome City Smoking  \
0          1   24    Male             93          169  Positive    B     nan   
1          2   61  Female            100          280  Negative    A     Yes   
2          3   69  Female             95          234  Positive    C      No   
3          4   52    Male            107          214  Negative    C      No   
4          5   58    Male            159          237  Positive    B      No   

         BMI Insurance  
0  27.983923        No  
1  27.637276       Yes  
2  28.541070        No  
3  23.418728       Yes  
4  34.975552       Yes

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 10 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PatientID      100 non-null    int64  
 1   Age            100 non-null    int64  
 2   Gender         100 non-null    object 
 3   BloodPressure  100 non-null    int64  
 4   Cholesterol    100 non-null    int64  
 5   Outcome        100 non-null    object 
 6   City           100 non-null    object 
 7   Smoking        100 non-null    object 
 8   BMI            100 non-null    float64
 9   Insurance      100 non-null    object 
dtypes: float64(1), int64(4), object(5)
memory usage: 7.9+ KB


None

PatientID        0
Age              0
Gender           0
BloodPressure    0
Cholesterol      0
Outcome          0
City             0
Smoking          0
BMI              0
Insurance        0
dtype: int64

PatientID        0.0
Age              0.0
Gender           0.0
BloodPressure    0.0
Cholesterol      0.0
Outcome          0.0
City             0.0
Smoking          0.0
BMI              0.0
Insurance        0.0
dtype: float64

## Handle missing values

### Subtask:
Handle missing values in the dataset.


## Encode categorical features

### Subtask:
Encode categorical features in the dataset.


**Reasoning**:
Identify categorical columns, determine their type, and apply appropriate encoding techniques. Then display the encoded DataFrame and its data types.



In [19]:
categorical_cols = df.select_dtypes(include='object').columns
print("Categorical columns:", categorical_cols.tolist())

nominal_cols = ['Gender', 'City', 'Smoking', 'Insurance']
ordinal_cols = ['Outcome']

df_encoded = pd.get_dummies(df, columns=nominal_cols, drop_first=True)

outcome_mapping = {'Negative': 0, 'Positive': 1}
df_encoded['Outcome'] = df_encoded['Outcome'].map(outcome_mapping)

display(df_encoded.head())

display(df_encoded.info())

Categorical columns: ['Gender', 'Outcome', 'City', 'Smoking', 'Insurance']


PatientID  Age  BloodPressure  Cholesterol  Outcome        BMI  \
0          1   24             93          169        1  27.983923   
1          2   61            100          280        0  27.637276   
2          3   69             95          234        1  28.541070   
3          4   52            107          214        0  23.418728   
4          5   58            159          237        1  34.975552   

   Gender_Male  City_B  City_C  City_D  Smoking_Yes  Smoking_nan  \
0         True    True   False   False        False         True   
1        False   False   False   False         True        False   
2        False   False    True   False        False        False   
3         True   False    True   False        False        False   
4         True    True   False   False        False        False   

   Insurance_Yes  
0          False  
1           True  
2          False  
3           True  
4           True

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100 entries, 0 to 99
Data columns (total 13 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   PatientID      100 non-null    int64  
 1   Age            100 non-null    int64  
 2   BloodPressure  100 non-null    int64  
 3   Cholesterol    100 non-null    int64  
 4   Outcome        100 non-null    int64  
 5   BMI            100 non-null    float64
 6   Gender_Male    100 non-null    bool   
 7   City_B         100 non-null    bool   
 8   City_C         100 non-null    bool   
 9   City_D         100 non-null    bool   
 10  Smoking_Yes    100 non-null    bool   
 11  Smoking_nan    100 non-null    bool   
 12  Insurance_Yes  100 non-null    bool   
dtypes: bool(7), float64(1), int64(5)
memory usage: 5.5 KB


None

## Split data

### Subtask:
Split the data into training and testing sets.


**Reasoning**:
Split the encoded data into features (X) and target (y), then split these into training and testing sets.



In [20]:
from sklearn.model_selection import train_test_split

X = df_encoded.drop(['Outcome', 'PatientID'], axis=1)

y = df_encoded['Outcome']

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

print("Shape of X_train:", X_train.shape)
print("Shape of X_test:", X_test.shape)
print("Shape of y_train:", y_train.shape)
print("Shape of y_test:", y_test.shape)

Shape of X_train: (80, 11)
Shape of X_test: (20, 11)
Shape of y_train: (80,)
Shape of y_test: (20,)


## Train decision tree model

### Subtask:
Train a Decision Tree Classifier on the preprocessed training data.


**Reasoning**:
Import the Decision Tree Classifier and train the model using the training data.



In [21]:
from sklearn.tree import DecisionTreeClassifier

clf = DecisionTreeClassifier(random_state=42)

clf.fit(X_train, y_train)

DecisionTreeClassifier(random_state=42)

## Tune hyperparameters

### Subtask:
Use techniques like GridSearchCV to find the optimal hyperparameters for the Decision Tree model.


**Reasoning**:
Use GridSearchCV to find the best hyperparameters for the Decision Tree Classifier using the training data.



In [16]:
from sklearn.model_selection import GridSearchCV

param_grid = {
    'max_depth': [None, 2, 3, 4, 5, 10, 15, 20],
    'min_samples_split': [2, 5, 10, 20, 30, 40],
    'min_samples_leaf': [1, 5, 10, 20]
}

clf = DecisionTreeClassifier(random_state=42)

grid_search = GridSearchCV(clf, param_grid, cv=5, scoring='accuracy')

grid_search.fit(X_train, y_train)

print("Best parameters found by GridSearchCV:")
print(grid_search.best_params_)

best_clf = grid_search.best_estimator_

Best parameters found by GridSearchCV:
{'max_depth': 5, 'min_samples_leaf': 5, 'min_samples_split': 20}


## Evaluate model performance

### Subtask:
Evaluate the performance of the tuned Decision Tree model using appropriate metrics for classification.


**Reasoning**:
Evaluate the performance of the tuned Decision Tree model using appropriate classification metrics.



In [22]:
from sklearn.metrics import accuracy_score, classification_report

y_pred_tuned = best_clf.predict(X_test)

accuracy_tuned = accuracy_score(y_test, y_pred_tuned)

report = classification_report(y_test, y_pred_tuned)

print(f"Accuracy of the tuned model: {accuracy_tuned:.2f}")
print("\nClassification Report of the tuned model:")
print(report)

Accuracy of the tuned model: 0.35

Classification Report of the tuned model:
              precision    recall  f1-score   support

           0       0.33      0.30      0.32        10
           1       0.36      0.40      0.38        10

    accuracy                           0.35        20
   macro avg       0.35      0.35      0.35        20
weighted avg       0.35      0.35      0.35        20



## Describe business value

### Subtask:
Describe the real-world business value of this model in a healthcare setting.


## Summary:

### Data Analysis Key Findings

*   The dataset initially appeared to have missing values based on the task description, but the initial exploration revealed no missing data in any of the columns.
*   The dataset contains a mix of data types, including numerical (`int64`, `float64`) and categorical (`object`).
*   Categorical features were successfully identified and encoded using one-hot encoding for nominal features (Gender, City, Smoking, Insurance) and ordinal encoding for the target variable (Outcome).
*   The dataset was split into training and testing sets with an 80/20 ratio.
*   Hyperparameter tuning using `GridSearchCV` identified the best parameters for the Decision Tree model as `max_depth: 5`, `min_samples_leaf: 5`, and `min_samples_split: 20`.
*   The performance of the tuned Decision Tree model on the test set was low, with an accuracy of 0.35 and F1-scores of 0.32 for class 0 and 0.38 for class 1.

### Insights or Next Steps

*   The low performance metrics suggest that a Decision Tree model may not be the most suitable model for this dataset and prediction task. Further exploration of other classification algorithms (e.g., Logistic Regression, Support Vector Machines, or ensemble methods) is recommended.
*   Given the small dataset size (100 entries), the model's performance might be limited by the amount of available data. Gathering more data or employing techniques like cross-validation during training could potentially improve performance.
